File: lib/bin/main/dbassistant/tokenhandlers/GenericTokenHandler.kt
package dbassistant.tokenhandlers

import dbassistant.exceptions.ImmutableHandlerException
import dbassistant.exceptions.MishandledTokenException
import dbassistant.interfaces.ITokenHandler
import dbassistant.tokens.Token
import dbassistant.tokens.TokenType

class GenericTokenHandler(
    override val priority: Int,
    maxGenericTokenCount: Int
) : ITokenHandler{
    
    private val words = MutableList(maxGenericTokenCount){""}
    
    override val size: Int get() = words.size

    override var mutable: Boolean = true

    override fun canEncode(word: String): Boolean {
        return true
    }
    
    override fun encode(word: String): Token {
        val index = words.indexOf(word)
        if(index == -1){
            if(mutable){
                val id = words.indexOfFirst { w -> w == "" }
                words[id] = word
                return Token(id,TokenType.GENERIC)
            }
            else{
                throw ImmutableHandlerException()
            }
        }
        return Token(index,TokenType.GENERIC)
    }

    override fun decode(token: Int): String {
        try{
            val word = words.elementAt(token)
            return word
        }
        catch (e:IndexOutOfBoundsException){
            throw MishandledTokenException()
        }
    }
    
    override fun resetState() {}

}

File: dbassistant/tokenhandlers/GenericTokenHandler.py
from dbassistant.exceptions.ArgumentHandlerIsFullException import ArgumentHandlerIsFullException
from dbassistant.exceptions.ImmutableHandlerException import ImmutableHandlerException
from dbassistant.exceptions.MishandledTokenException import MishandledTokenException
from dbassistant.interfaces.ITokenHandler import ITokenHandler
from dbassistant.tokens.Token import Token
from dbassistant.tokens.TokenType import TokenType


class GenericTokenHandler(ITokenHandler):
    def __init__(self, priority: int, max_generic_token_count: int):
        self._priority = priority
        self.words = [""] * max_generic_token_count
        self._mutable = True

    @property
    def priority(self):
        return self._priority
    
    @priority.setter
    def priority(self, value):
        self._priority = value
    
    @property
    def learned_tokens(self):
        return self.words.copy()

    @property
    def size(self):
        return len(self.words)

    @property
    def utilizedSize(self):
        return len([w for w in self.words if w != ""])

    @property
    def mutable(self):
        return self._mutable

    def canEncode(self, word: str):
        return True

    def encode(self, word: str):
        index = self.words.index(word) if word in self.words else -1
        if index == -1:
            if self.mutable:
                try:
                    id = self.words.index("")
                    self.words[id] = word
                    return Token(id, TokenType.GENERIC)
                except ValueError:
                    pass
            else:
                raise ImmutableHandlerException()
        return Token(index, TokenType.GENERIC)

    def decode(self, token: int):
        try:
            return self.words[token]
        except IndexError:
            raise MishandledTokenException()

    def resetState(self):
        pass

    def serialize(self, path):
        file = path + "/generic_tokens.layer"
        with open(file, "w") as f:
            for word in self.words:
                f.write(word + "\n")

    def deserialize(self,path):
        file = path + "/generic_tokens.layer"
        with open(file, "r") as f:
            for line in f:
                self.encode(line.strip())
        self._mutable = False
